src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs

```diff
@@ -2,214 +2,151 @@
+        // * Dynamic table encoding:
+        //   This would make the encoder stateful, which complicates things significantly.
+        //   Additionally, it's not clear exactly what strings we would add to the dynamic table
+        //   without some additional guidance from the user about this.
+        //   So for now, don't do dynamic encoding.
 
-        public static bool EncodeIndexedField(int index, Span<byte> buffer, out int length)
+        public static bool EncodeIndexedField(int index, Span<byte> destination, out int bytesWritten)
         {
-            length = 0;
-
-            if (buffer.Length == 0)
-            {
-                return false;
-            }
-
-            buffer[0] = 0x80;   // Literal field
-            if (!IntegerEncoder.Encode(index, 7, buffer, out length))
+            if (destination.Length != 0)
             {
-                return false;
+                destination[0] = 0x80;   // Literal field
+                return IntegerEncoder.Encode(index, 7, destination, out bytesWritten);
             }
 
-            return true;
+            bytesWritten = 0;
+            return false;
         }
 
-        public static bool EncodeIndexedName(int index, string value, Span<byte> buffer, out int length)
+        public static bool EncodeIndexedName(int index, string value, Span<byte> destination, out int bytesWritten)
         {
-            int i = 0;
-            length = 0;
-
-            if (buffer.Length == 0)
+            if (destination.Length != 0)
             {
-                return false;
-            }
-
-            buffer[0] = 0x40;   // Literal name
-            if (!IntegerEncoder.Encode(index, 6, buffer, out int indexLength))
-            {
-                return false;
-            }
-
-            i = indexLength;
-            if (i == buffer.Length)
-            {
-                return false;
-            }
-
-            if (!EncodeString(value, buffer.Slice(i), out int nameLength, lowercase: false))
-            {
-                return false;
+                destination[0] = 0x40;   // Literal name
+                if (IntegerEncoder.Encode(index, 6, destination, out int indexLength))
+                {
+                    if (indexLength < destination.Length &&
+                        EncodeString(value, destination.Slice(indexLength), out int nameLength, toLower: false))
+                    {
+                        bytesWritten = indexLength + nameLength;
+                        return true;
+                    }
+                }
             }
 
-            i += nameLength;
-
-            length = i;
-            return true;
+            bytesWritten = 0;
+            return false;
         }
 
-        public static bool EncodeIndexedName(int index, ReadOnlySpan<byte> value, Span<byte> buffer, out int length)
+        public static bool EncodeIndexedName(int index, ReadOnlySpan<byte> value, Span<byte> destination, out int bytesWritten)
```

Comment:
I think this overload is only used for the cached Host header, right?

I wonder if we should just eliminate this and pass the Host header as a string to the above routine. That way we have less code. This only happens once per HttpConnectionPool so it doesn't seem like a perf issue.

This would also simplify moving to Huffman encoding in the future for this value.

Reply:
This also allows us to remove EncodeAsciiString since this is the only place it's used.

Reply:
Ok